apps/src/feedback.js

```diff
@@ -453,28 +453,55 @@ FeedbackUtils.prototype.displayFeedback = function (options, requiredBlocks,
     });
   }
 
-  // set up the Save To Gallery button if necessary
-  var saveToGalleryButton = feedback.querySelector('#save-to-gallery-button');
-  if (saveToGalleryButton && options.saveToProjectGallery) {
-    dom.addClickTouchEvent(saveToGalleryButton, () => {
-      $('#save-to-gallery-button').prop('disabled', true).text("Saving...");
-      project.copy(project.getNewProjectName(), () => {
-        dataURIToBlob(options.feedbackImage)
-          .then(project.saveThumbnail)
-          .then(() => {
-            return new Promise(resolve => {
-              // Save the thumbnail url to the server.
-              project.save(resolve);
-            });
-          }).then(() => {
-            $('#save-to-gallery-button').prop('disabled', true).text("Saved!");
-          });
-      }, {shouldPublish: true});
-    });
-  } else if (saveToGalleryButton && options.response && options.response.save_to_gallery_url) {
-    dom.addClickTouchEvent(saveToGalleryButton, function () {
-      $.post(options.response.save_to_gallery_url,
-             function () { $('#save-to-gallery-button').prop('disabled', true).text("Saved!"); });
+  /**
+   * Add an event handler to the specified button which saves the project
+   * with thumbnail to the project gallery, and optionally publishes it.
+   * @param {string} buttonSelector Selector for the button html element.
+   * @param {string} pendingText Button text while operation is pending.
+   * @param {string} finishedText Button text after operation completes.
+   * @param {boolean} shouldPublish Whether to publish. Default: false.
+   * @param {function} callback Callback function to call on success.
+   */
+  function addSaveProjectButtonHandler(buttonSelector, pendingText, finishedText, shouldPublish, callback) {
+    const buttonElement = feedback.querySelector(buttonSelector);
+    if (buttonElement) {
+      dom.addClickTouchEvent(buttonElement, () => {
+        $(buttonSelector).prop('disabled', true).text(pendingText);
+        project.copy(project.getNewProjectName(), () => {
+          FeedbackUtils.saveThumbnail(options.feedbackImage).then(() => {
+            $(buttonSelector).prop('disabled', true).text(finishedText);
+          }).then(callback);
+        }, {shouldPublish});
+      });
+    }
+  }
+
+  const saveButtonSelector = '#save-to-project-gallery-button';
+  addSaveProjectButtonHandler(
+    saveButtonSelector,
+    msg.saving(),
+    msg.savedToGallery()
+  );
+
+  const publishButtonSelector = '#publish-to-project-gallery-button';
+  addSaveProjectButtonHandler(
+    publishButtonSelector,
+    msg.publishPending(),
+    msg.published(),
+    true,
+    () => {
+      // Publishing the project also saves it, so disable the save button.
+      $(saveButtonSelector).prop('disabled', true).text(msg.savedToGallery());
```

Comment:
I started to do this, but then I realized concluded that it doesn't really make sense to disable only the save button while the publish is pending, because really we don't want people to click on any other buttons during that time. But once the publish request completes, it does make sense to disable the Save button, because at that point we know the project has been saved so an additional Save is not necessary.